File: ai_service/app/services/search_service.py
"""
Search service for semantic and keyword-based document search
"""

import time
from typing import List, Optional
from app.models.search import SearchRequest, SearchResponse, SearchResult, SearchMode
from app.database.vector_db import VectorDatabase
from app.services.embedding_service import EmbeddingService
from app.utils.logger import logger


class SearchService:
    """Service for searching documents"""

    def __init__(self, user_id: str, embedding_service: Optional[EmbeddingService] = None):
        """
        Initialize search service

        Args:
            user_id: User identifier
            embedding_service: Optional embedding service instance
        """
        self.user_id = user_id
        self.vector_db = VectorDatabase(user_id)
        self.embedding_service = embedding_service or EmbeddingService()

    async def search(self, request: SearchRequest) -> SearchResponse:
        """
        Perform search based on request parameters

        Args:
            request: SearchRequest object

        Returns:
            SearchResponse with results
        """
        start_time = time.time()

        # Route to appropriate search method
        if request.mode == SearchMode.SEMANTIC:
            results = await self._semantic_search(request)
        elif request.mode == SearchMode.KEYWORD:
            results = await self._keyword_search(request)
        elif request.mode == SearchMode.HYBRID:
            results = await self._hybrid_search(request)
        else:
            raise ValueError(f"Unsupported search mode: {request.mode}")

        processing_time = (time.time() - start_time) * 1000  # Convert to ms

        logger.info(
            f"Search completed: mode={request.mode}, "
            f"results={len(results)}, time={processing_time:.2f}ms"
        )

        return SearchResponse(
            results=results,
            total=len(results),
            query=request.query,
            mode=request.mode,
            processing_time_ms=processing_time
        )

    async def _semantic_search(self, request: SearchRequest) -> List[SearchResult]:
        """
        Perform semantic (vector similarity) search

        Args:
            request: SearchRequest object

        Returns:
            List of SearchResult objects
        """
        # Generate query embedding
        query_embedding = await self.embedding_service.generate_query_embedding(
            request.query,
            provider="openai"  # Can be made configurable
        )

        # Perform vector search
        results = await self.vector_db.search_by_vector(
            query_embedding=query_embedding,
            top_k=request.top_k,
            document_id=request.document_id,
            tags=request.tags,
            min_score=request.min_score
        )

        return results

    async def _keyword_search(self, request: SearchRequest) -> List[SearchResult]:
        """
        Perform keyword-based text search

        Args:
            request: SearchRequest object

        Returns:
            List of SearchResult objects
        """
        # Generate query embedding to ensure dimension consistency
        query_embedding = await self.embedding_service.generate_query_embedding(
            request.query,
            provider="openai"
        )
        
        results = await self.vector_db.search_by_keyword(
            query=request.query,
            top_k=request.top_k,
            document_id=request.document_id,
            tags=request.tags,
            query_embedding=query_embedding
        )

        return results

    async def _hybrid_search(self, request: SearchRequest) -> List[SearchResult]:
        """
        Perform hybrid search combining semantic and keyword search

        Args:
            request: SearchRequest object

        Returns:
            List of SearchResult objects (merged and re-ranked)
        """
        # Perform both searches
        semantic_results = await self._semantic_search(request)
        keyword_results = await self._keyword_search(request)

        # Merge results and re-rank
        merged_results = self._merge_and_rerank(
            semantic_results,
            keyword_results,
            request.top_k
        )

        return merged_results

    @staticmethod
    def _merge_and_rerank(
        semantic_results: List[SearchResult],
        keyword_results: List[SearchResult],
        top_k: int,
        semantic_weight: float = 0.7,
        keyword_weight: float = 0.3
    ) -> List[SearchResult]:
        """
        Merge and re-rank results from semantic and keyword search

        Args:
            semantic_results: Results from semantic search
            keyword_results: Results from keyword search
            top_k: Number of results to return
            semantic_weight: Weight for semantic scores
            keyword_weight: Weight for keyword scores

        Returns:
            Merged and re-ranked results
        """
        # Create a dictionary to store combined scores
        result_map = {}

        # Add semantic results
        for result in semantic_results:
            result_map[result.chunk_id] = {
                "result": result,
                "semantic_score": result.score,
                "keyword_score": 0.0
            }

        # Add/update with keyword results
        for result in keyword_results:
            if result.chunk_id in result_map:
                result_map[result.chunk_id]["keyword_score"] = result.score
            else:
                result_map[result.chunk_id] = {
                    "result": result,
                    "semantic_score": 0.0,
                    "keyword_score": result.score
                }

        # Calculate combined scores
        ranked_results = []
        for chunk_id, data in result_map.items():
            combined_score = (
                semantic_weight * data["semantic_score"] +
                keyword_weight * data["keyword_score"]
            )

            result = data["result"]
            result.score = combined_score
            ranked_results.append(result)

        # Sort by combined score
        ranked_results.sort(key=lambda x: x.score, reverse=True)

        return ranked_results[:top_k]

    async def search_similar_chunks(
        self,
        document_id: str,
        chunk_index: int,
        top_k: int = 5
    ) -> List[SearchResult]:
        """
        Find similar chunks to a specific chunk

        Args:
            document_id: Document identifier
            chunk_index: Index of the chunk
            top_k: Number of similar chunks to return

        Returns:
            List of similar SearchResult objects
        """
        # Get the reference chunk
        chunks = await self.vector_db.get_document_chunks(
            document_id,
            include_embeddings=True
        )

        reference_chunk = None
        for chunk in chunks:
            if chunk["chunk_index"] == chunk_index:
                reference_chunk = chunk
                break

        if not reference_chunk or not reference_chunk.get("embedding"):
            logger.warning(f"Chunk not found or has no embedding: {document_id}:{chunk_index}")
            return []

        # Search using the chunk's embedding
        results = await self.vector_db.search_by_vector(
            query_embedding=reference_chunk["embedding"],
            top_k=top_k + 1  # +1 because it will include itself
        )

        # Filter out the reference chunk itself
        results = [r for r in results if r.chunk_id != reference_chunk["chunk_id"]]

        return results[:top_k]

File: src/lib/document/search-service.ts
/**
 * Document Search Service
 *
 * Service for searching documents using vector similarity and keyword search.
 */

import {
  type SearchQuery,
  type SearchResponse,
  type SearchResult,
  SearchMode,
  type DocumentMetadata,
  type DocumentChunk,
} from '@/types/document';

import {
  getDocumentsCollection,
  getUserVectorCollection,
} from '@/lib/db/document-db';

import { EmbeddingService } from './embedding-service';
import { Filter } from 'mongodb';

// ============================================================================
// Search Service Class
// ============================================================================

export class SearchService {
  private embeddingService: EmbeddingService;

  constructor() {
    this.embeddingService = new EmbeddingService();
  }

  /**
   * Search documents
   */
  async search(query: SearchQuery): Promise<SearchResponse> {
    const startTime = Date.now();

    try {
      // Validate user has documents
      const vectorCollection = await getUserVectorCollection(query.userId);
      const hasDocuments = (await vectorCollection.estimatedDocumentCount()) > 0;

      if (!hasDocuments) {
        return {
          results: [],
          totalResults: 0,
          query: query.query,
          searchMode: query.mode,
          processingTime: Date.now() - startTime,
          hasMore: false,
        };
      }

      // Execute search based on mode
      let results: SearchResult[] = [];

      switch (query.mode) {
        case SearchMode.SEMANTIC:
          results = await this.semanticSearch(query);
          break;

        case SearchMode.KEYWORD:
          results = await this.keywordSearch(query);
          break;

        case SearchMode.HYBRID:
          results = await this.hybridSearch(query);
          break;

        case SearchMode.VISUAL:
          results = await this.visualSearch(query);
          break;

        default:
          throw new Error(`Unsupported search mode: ${query.mode}`);
      }

      // Apply post-processing filters
      results = this.applyPostFilters(results, query);

      // Rerank if requested
      if (query.rerank && results.length > 0) {
        results = await this.rerankResults(results, query.query);
      }

      // Apply pagination
      const offset = query.offset || 0;
      const limit = query.limit || 10;
      const paginatedResults = results.slice(offset, offset + limit);

      const processingTime = Date.now() - startTime;

      return {
        results: paginatedResults,
        totalResults: results.length,
        query: query.query,
        searchMode: query.mode,
        processingTime,
        hasMore: offset + limit < results.length,
        nextOffset: offset + limit < results.length ? offset + limit : undefined,
      };
    } catch (error: any) {
      console.error('Search error:', error);
      throw new Error(`Search failed: ${error.message}`);
    }
  }

  /**
   * Semantic search using vector similarity
   */
  private async semanticSearch(query: SearchQuery): Promise<SearchResult[]> {
    // Generate query embedding
    const embeddingResult = await this.embeddingService.generateEmbedding({
      text: query.query,
    });

    const queryEmbedding = Array.isArray(embeddingResult.embeddings[0])
      ? embeddingResult.embeddings[0]
      : embeddingResult.embeddings;

    // Build MongoDB aggregation pipeline for vector search
    const vectorCollection = await getUserVectorCollection(query.userId);

    // Build filter
    const filter = this.buildFilter(query);

    // MongoDB Atlas Vector Search (requires Atlas Search index)
    // Note: This requires MongoDB Atlas with vector search capability
    const pipeline: any[] = [
      {
        $vectorSearch: {
          index: 'vector_index', // Name of the vector search index
          path: 'embedding',
          queryVector: queryEmbedding,
          numCandidates: (query.topK || 10) * 10, // Search more candidates
          limit: query.topK || 10,
          filter: filter,
        },
      },
      {
        $addFields: {
          score: { $meta: 'vectorSearchScore' },
        },
      },
      {
        $match: {
          score: { $gte: query.minScore || 0.5 },
        },
      },
    ];

    try {
      const chunks = await vectorCollection.aggregate(pipeline).toArray();

      // Convert to search results
      return await this.chunksToSearchResults(
        chunks as DocumentChunk[],
        query,
        'semantic'
      );
    } catch (error: any) {
      // Fallback to cosine similarity if vector search is not available
      console.warn('Vector search not available, using fallback:', error.message);
      return this.fallbackSemanticSearch(query, queryEmbedding as number[]);
    }
  }

  /**
   * Fallback semantic search using manual cosine similarity
   */
  private async fallbackSemanticSearch(
    query: SearchQuery,
    queryEmbedding: number[]
  ): Promise<SearchResult[]> {
    const vectorCollection = await getUserVectorCollection(query.userId);
    const filter = this.buildFilter(query);

    // Fetch all chunks (or a subset)
    const chunks = await vectorCollection
      .find(filter)
      .limit(1000) // Limit to prevent memory issues
      .toArray();

    // Calculate cosine similarity for each chunk
    const scoredChunks = chunks.map(chunk => ({
      ...chunk,
      score: this.cosineSimilarity(queryEmbedding, chunk.embedding),
    }));

    // Sort by score and filter
    const filteredChunks = scoredChunks
      .filter(chunk => chunk.score >= (query.minScore || 0.5))
      .sort((a, b) => b.score - a.score)
      .slice(0, query.topK || 10);

    return this.chunksToSearchResults(filteredChunks, query, 'semantic');
  }

  /**
   * Keyword search using text search
   */
  private async keywordSearch(query: SearchQuery): Promise<SearchResult[]> {
    const vectorCollection = await getUserVectorCollection(query.userId);
    const filter = this.buildFilter(query);

    // Add text search to filter
    const searchFilter = {
      ...filter,
      $text: { $search: query.query },
    };

    try {
      // MongoDB text search
      const chunks = await vectorCollection
        .find(searchFilter, {
          score: { $meta: 'textScore' },
        })
        .sort({ score: { $meta: 'textScore' } })
        .limit(query.topK || 10)
        .toArray();

      return this.chunksToSearchResults(chunks as DocumentChunk[], query, 'keyword');
    } catch (error: any) {
      // Fallback to regex search if text index doesn't exist
      console.warn('Text search not available, using regex fallback');
      return this.fallbackKeywordSearch(query);
    }
  }

  /**
   * Fallback keyword search using regex
   */
  private async fallbackKeywordSearch(query: SearchQuery): Promise<SearchResult[]> {
    const vectorCollection = await getUserVectorCollection(query.userId);
    const filter = this.buildFilter(query);

    const searchFilter = {
      ...filter,
      content: { $regex: query.query, $options: 'i' },
    };

    const chunks = await vectorCollection
      .find(searchFilter)
      .limit(query.topK || 10)
      .toArray();

    return this.chunksToSearchResults(chunks as DocumentChunk[], query, 'keyword');
  }

  /**
   * Hybrid search (combines semantic and keyword)
   */
  private async hybridSearch(query: SearchQuery): Promise<SearchResult[]> {
    // Run both searches in parallel
    const [semanticResults, keywordResults] = await Promise.all([
      this.semanticSearch(query),
      this.keywordSearch(query),
    ]);

    // Merge and deduplicate results using reciprocal rank fusion
    const mergedResults = this.reciprocalRankFusion(
      semanticResults,
      keywordResults,
      query.topK || 10
    );

    return mergedResults;
  }

  /**
   * Visual search (for images)
   */
  private async visualSearch(query: SearchQuery): Promise<SearchResult[]> {
    // TODO: Implement visual similarity search
    // This would require visual embeddings (e.g., CLIP)
    throw new Error('Visual search not yet implemented');
  }

  /**
   * Build MongoDB filter from query
   */
  private buildFilter(query: SearchQuery): Filter<DocumentChunk> {
    const filter: Filter<DocumentChunk> = {
      userId: query.userId,
    };

    if (query.trackingIds && query.trackingIds.length > 0) {
      filter.trackingId = { $in: query.trackingIds };
    }

    if (query.tags && query.tags.length > 0) {
      filter.tags = { $in: query.tags };
    }

    if (query.dateRange) {
      filter.createdAt = {};
      if (query.dateRange.from) {
        filter.createdAt.$gte = query.dateRange.from;
      }
      if (query.dateRange.to) {
        filter.createdAt.$lte = query.dateRange.to;
      }
    }

    return filter;
  }

  /**
   * Convert chunks to search results
   */
  private async chunksToSearchResults(
    chunks: any[],
    query: SearchQuery,
    searchType: string
  ): Promise<SearchResult[]> {
    // Get document metadata if requested
    let documentsMap = new Map<string, DocumentMetadata>();

    if (query.includeMetadata) {
      const trackingIds = [...new Set(chunks.map(c => c.trackingId))];
      const documentsCollection = await getDocumentsCollection();
      const documents = await documentsCollection
        .find({ trackingId: { $in: trackingIds } })
        .toArray();

      documents.forEach(doc => {
        documentsMap.set(doc.trackingId, doc);
      });
    }

    // Build search results
    return chunks.map(chunk => {
      const result: SearchResult = {
        chunkId: chunk.chunkId,
        trackingId: chunk.trackingId,
        content: query.includeContent ? chunk.content : '',
        score: chunk.score || 1.0,
        metadata: chunk.metadata,
      };

      if (query.includeMetadata && documentsMap.has(chunk.trackingId)) {
        const docMetadata = documentsMap.get(chunk.trackingId)!;
        result.documentTitle = docMetadata.title || docMetadata.fileName;
        result.documentMetadata = {
          fileName: docMetadata.fileName,
          documentType: docMetadata.documentType,
          uploadDate: docMetadata.uploadDate,
          tags: docMetadata.tags,
        };
      }

      // Generate highlights for keyword search
      if (searchType === 'keyword') {
        result.highlights = this.generateHighlights(chunk.content, query.query);
      }

      return result;
    });
  }

  /**
   * Apply post-processing filters
   */
  private applyPostFilters(
    results: SearchResult[],
    query: SearchQuery
  ): SearchResult[] {
    return results.filter(result => result.score >= (query.minScore || 0.5));
  }

  /**
   * Rerank search results
   */
  private async rerankResults(
    results: SearchResult[],
    query: string
  ): Promise<SearchResult[]> {
    // TODO: Implement reranking using cross-encoder or similar
    // For now, just return as-is
    return results;
  }

  /**
   * Reciprocal rank fusion for combining results
   */
  private reciprocalRankFusion(
    list1: SearchResult[],
    list2: SearchResult[],
    k: number = 60
  ): SearchResult[] {
    const scores = new Map<string, { result: SearchResult; score: number }>();

    // Calculate RRF scores for list1
    list1.forEach((result, index) => {
      const rrfScore = 1 / (k + index + 1);
      scores.set(result.chunkId, { result, score: rrfScore });
    });

    // Add RRF scores for list2
    list2.forEach((result, index) => {
      const rrfScore = 1 / (k + index + 1);
      const existing = scores.get(result.chunkId);

      if (existing) {
        existing.score += rrfScore;
      } else {
        scores.set(result.chunkId, { result, score: rrfScore });
      }
    });

    // Sort by combined score
    return Array.from(scores.values())
      .sort((a, b) => b.score - a.score)
      .map(item => ({
        ...item.result,
        score: item.score,
      }));
  }

  /**
   * Calculate cosine similarity
   */
  private cosineSimilarity(a: number[], b: number[]): number {
    if (a.length !== b.length) {
      throw new Error('Vectors must have the same length');
    }

    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < a.length; i++) {
      dotProduct += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }

    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  /**
   * Generate text highlights
   */
  private generateHighlights(text: string, query: string): string[] {
    const highlights: string[] = [];
    const words = query.toLowerCase().split(/\s+/);
    const sentences = text.split(/[.!?]+/);

    for (const sentence of sentences) {
      const lowerSentence = sentence.toLowerCase();
      if (words.some(word => lowerSentence.includes(word))) {
        highlights.push(sentence.trim());
        if (highlights.length >= 3) break;
      }
    }

    return highlights;
  }
}

// ============================================================================
// Export Default Instance
// ============================================================================

/**
 * Default search service instance
 */
export const defaultSearchService = new SearchService();
